### PR TITLE
gh-129061: Fix `FORCE_COLOR` and `NO_COLOR` when empty strings

### PR DESCRIPTION
**Pull request by hugovk**
First refactor the tests into subtests, which makes it easier compare the env var inputs and expected outputs.

* Issue: gh-129061

### CODE REVIEW

**Comment by hugovk** [comment]
!buildbot ARM64 Windows Non-Debug

**Comment by hugovk** [comment]
!buildbot AMD64

**Comment by hugovk** [comment]
!buildbot aarch64

**Comment by hugovk** [comment]
!buildbot s390x RHEL8 Refleaks

**Comment by hugovk** [comment]
!buildbot s390x RHEL9 Refleaks

**Comment by hugovk** [comment]
Like https://github.com/python/cpython/pull/129137#issuecomment-2608701118, `test_external_inspection` failed on Fedora and RHEL8 buildbots and is unrelated.

File: Lib/test/test__colorize.py [hunk 1]
```diff
@@ -42,28 +42,40 @@ def test_colorized_detection_checks_for_environment_variables(self):
...
-                                     {'FORCE_COLOR': '1', 'NO_COLOR': '1'}):
-                self.assertEqual(_colorize.can_colorize(), False)
-            with unittest.mock.patch("os.environ",
-                                     {'FORCE_COLOR': '1', "PYTHON_COLORS": '0'}):
-                self.assertEqual(_colorize.can_colorize(), False)
+
+            for env_vars, expected in [
+                ({"TERM": "dumb"}, False),
+                ({"PYTHON_COLORS": "1"}, True),
+                ({"PYTHON_COLORS": "0"}, False),
```
Comment (serhiy-storchaka) [inline Lib/test/test__colorize.py]: Test also an empty value. Or other values (e.g. "2").

File: Lib/test/test__colorize.py [hunk 2]
```diff
@@ -42,28 +42,40 @@ def test_colorized_detection_checks_for_environment_variables(self):
...
+                ({"NO_COLOR": "1"}, False),
+                ({"NO_COLOR": "0"}, False),
```
Comment (serhiy-storchaka) [inline Lib/test/test__colorize.py]: Test also an empty value.

File: Lib/test/test__colorize.py [hunk 3]
```diff
@@ -42,28 +42,40 @@ def test_colorized_detection_checks_for_environment_variables(self):
...
+                ({"NO_COLOR": "1", "PYTHON_COLORS": "1"}, True),
+                ({"FORCE_COLOR": "1"}, True),
```
Comment (serhiy-storchaka) [inline Lib/test/test__colorize.py]: Since the fallback value (determined by TERM, isatty or _supports_virtual_terminal) is True, setting FORCE_COLOR to "1" does not have any effect. We need to test it when the fallback value is False.

You can test `{"TERM": "dumb", "FORCE_COLOR": "1"}`.

File: Lib/test/test__colorize.py [hunk 4]
```diff
@@ -42,28 +42,40 @@ def test_colorized_detection_checks_for_environment_variables(self):
...
+                ({"FORCE_COLOR": "1", "NO_COLOR": "1"}, False),
+                ({"FORCE_COLOR": "1", "NO_COLOR": "0"}, False),
+                ({"FORCE_COLOR": "1", "NO_COLOR": ""}, True),
+                ({"FORCE_COLOR": "0", "NO_COLOR": "1"}, False),
```
Comment (serhiy-storchaka) [inline Lib/test/test__colorize.py]: Both variables set the same condition. This test does not add anything.

File: Lib/test/test__colorize.py [hunk 5]
```diff
@@ -42,28 +42,40 @@ def test_colorized_detection_checks_for_environment_variables(self):
...
+                ({"FORCE_COLOR": "", "NO_COLOR": "1"}, False),
```
Comment (serhiy-storchaka) [inline Lib/test/test__colorize.py]: The FORCE_COLOR values does not matter, and this was already tested by `{"FORCE_COLOR": "1", "NO_COLOR": "1"}`. So this test is unneded.

**Review by serhiy-storchaka — COMMENTED** [review]
We only need to test pairs of environment variables that set conflicting values. 
For example, `NO_COLOR=1` and `TERM=dumb` should be tested with `FORCE_COLOR=1` and `PYTHON_COLOR=1`. Tests for a single environment variable which forces color (`FORCE_COLOR` and `PYTHON_COLOR`) should be performed when the fallback value determined by isatty or _supports_virtual_terminal is False.

We need also to test `can_colorize()` when stderr is a TTY, but stdout is not, and with an explicit file argument. Including corner cases: no `file.fileno`, no `file.isatty`, `file.isatty()` and `os.isatty(file.fileno())` return different results. BTW, I think that `file.isatty()` should perhaps be tested before `os.isatty(file.fileno())`.

This is a large issue and can be done in a separate PR.

File: Lib/test/test__colorize.py (hunk 4 above)
Comment (hugovk) [reply Lib/test/test__colorize.py]: These variables are contradictory, one says to force colour, the other to disable: 

* `"FORCE_COLOR": "0"`: the variable is present and not an empty string (regardless of its value) --> force colour
* `"NO_COLOR": "1"`: the variable is present and not an empty string (regardless of its value) --> disable colour
* But `NO_COLOR` takes priority over `FORCE_COLOR` --> disable colour wins

We're testing that `NO_COLOR` takes priority and that the value doesn't matter.

True, in the implementation, we return early when checking `NO_COLOR` before looking at `NO_COLOR`, but the tests don't care about implementation details, only the end result.

And these tests are fast to run.

File: Lib/test/test__colorize.py (hunk 3 above)
Comment (hugovk) [reply Lib/test/test__colorize.py]: `{"FORCE_COLOR": "1"}` returns early before checking `TERM`, `isatty` and `nt._supports_virtual_terminal`, so I think there is value testing `{"FORCE_COLOR": "1"}`.

I'll add `{"TERM": "dumb", "FORCE_COLOR": "1"}` as well.

File: Lib/test/test__colorize.py (hunk 5 above)
Comment (hugovk) [reply Lib/test/test__colorize.py]: * `"FORCE_COLOR": ""` is empty so this has no effect.
* `"NO_COLOR": "1"` means disable colour.
* So disabling colour wins.

Yes, this does have the same result as `{"FORCE_COLOR": "1", "NO_COLOR": "1"}`, and because we check `NO_COLOR` first and return early, `FORCE_COLOR` has no effect for both.

But I think it's fine to ensure these inputs give this expected output.

However, if you feel strongly about it, I'm happy to remove them.

File: Lib/test/test__colorize.py (hunk 1 above)
Comment (hugovk) [reply Lib/test/test__colorize.py]: Added "true" and "2".

**Review by serhiy-storchaka — APPROVED** [review]

**Comment by serhiy-storchaka** [comment]
See improved tests in #129234. They cover all corner cases. I propose to push it first, and then add tests for empty values.

**Review by serhiy-storchaka — COMMENTED** [review]
Only the first two checks are _necessary_. Others do not add anything useful. They look out of place and make the intention of the tests less clear.

The first group tests the different values of the variables. The second group tests relative priority of the variables (it does not repeat tests for different values). Then the same is done for PYTHON_COLORS -- first the different values of PYTHON_COLORS (including the empty value), then its relative priority to other variables.

**Review by serhiy-storchaka — APPROVED** [review]
LGTM! :+1:
